File: rust/src/leetcode/Q494_target_sum.rs
/*
 * @lc app=leetcode id=494 lang=rust
 *
 * [494] Target Sum
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        let mut possible_results: HashMap<i32, i32> = HashMap::new();
        possible_results.insert(0, 1);

        for num in nums {
            let mut new_results: HashMap<i32, i32> = HashMap::new();

            for (&total, &count) in &possible_results {
                *new_results.entry(total + num).or_insert(0) += count;
                *new_results.entry(total - num).or_insert(0) += count;
            }

            possible_results = new_results;
        }

        *possible_results.get(&target).unwrap_or(&0)
    }
}
// @lc code=end

File: python/Q494_target_sum.py
#
# @lc app=leetcode id=494 lang=python3
#
# [494] Target Sum
#


# @lc code=start
class Solution:
    def findTargetSumWays(self, nums: list[int], target: int) -> int:
        dp: list[dict[int, int]] = [dict() for _ in range(len(nums) + 1)]
        dp[0][0] = 1

        for i in range(1, len(nums) + 1):
            num: int = nums[i - 1]

            for total, count in dp[i - 1].items():
                dp[i][total + num] = dp[i].get(total + num, 0) + count
                dp[i][total - num] = dp[i].get(total - num, 0) + count

        return dp[-1].get(target, 0)


# @lc code=end
